File: notificare_push_ui/ios/Classes/NotificarePushUIPluginEventBroker.swift
//
//  NotificarePushUIPluginEventBroker.swift
//  notificare_push_ui
//
//  Created by Helder Pinhal on 06/06/2021.
//

import NotificareKit
import NotificarePushUIKit

class NotificarePushUIPluginEventBroker {
    private let namespace: String
    
    private var channels: [EventType: FlutterEventChannel] = [:]
    private var streams: [EventType: Stream]
    
    init(namespace: String) {
        var streams: [EventType: Stream] = [:]
        EventType.allCases.forEach { type in
            streams[type] = Stream(namespace: namespace, type: type)
        }
        
        self.namespace = namespace
        self.streams = streams
    }
    
    func setup(registrar: FlutterPluginRegistrar) {
        streams.values.forEach { stream in
            if let channel = channels[stream.type] {
                channel.setStreamHandler(stream)
            } else {
                let channel = FlutterEventChannel(
                    name: stream.name,
                    binaryMessenger: registrar.messenger(),
                    codec: FlutterJSONMethodCodec.sharedInstance()
                )
                
                channel.setStreamHandler(stream)

                channels[stream.type] = channel
            }
        }
    }
    
    func cleanup() {
        channels.values.forEach { channel in
            channel.setStreamHandler(nil)
        }
    }
    
    func emit(_ event: Event) {
        DispatchQueue.main.async { [weak self] in
            self?.streams[event.type]?.send(event)
        }
    }
}

// NotificarePushUIPluginEventBroker.Stream
extension NotificarePushUIPluginEventBroker {
    class Stream: NSObject, FlutterStreamHandler {
        let type: EventType
        let name: String

        private var eventSink: FlutterEventSink?
        private var pendingEvents: [Event] = []

        init(namespace: String, type: EventType) {
            self.type = type
            self.name = "\(namespace)/events/\(type.rawValue)"
        }

        func send(_ event: Event) {
            if let sink = self.eventSink {
                sink(event.payload)
            } else {
                pendingEvents.append(event)
            }
        }

        func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
            self.eventSink = events

            if (self.eventSink != nil) {
                self.pendingEvents.forEach { send($0) }
                self.pendingEvents.removeAll()
            }

            return nil
        }

        func onCancel(withArguments arguments: Any?) -> FlutterError? {
            self.eventSink = nil
            return nil
        }
    }
}

// NotificarePushUIPluginEventBroker.Event
extension NotificarePushUIPluginEventBroker {
    enum EventType: String, CaseIterable {
        case notificationWillPresent = "notification_will_present"
        case notificationPresented = "notification_presented"
        case notificationFinishedPresenting = "notification_finished_presenting"
        case notificationFailedToPresent = "notification_failed_to_present"
        case notificationUrlClicked = "notification_url_clicked"
        case actionWillExecute = "action_will_execute"
        case actionExecuted = "action_executed"
        case actionNotExecuted = "action_not_executed"
        case actionFailedToExecute = "action_failed_to_execute"
        case customActionReceived = "custom_action_received"
    }
    
    struct Event {
        let type: EventType
        let payload: Any?
    }
}

extension NotificarePushUIPluginEventBroker {
    static func OnNotificationWillPresent(notification: NotificareNotification) -> Event {
        return Event(
            type: .notificationWillPresent,
            payload: try! notification.toJson()
        )
    }
    
    static func OnNotificationPresented(notification: NotificareNotification) -> Event {
        return Event(
            type: .notificationPresented,
            payload: try! notification.toJson()
        )
    }
    
    static func OnNotificationFinishedPresenting(notification: NotificareNotification) -> Event {
        return Event(
            type: .notificationFinishedPresenting,
            payload: try! notification.toJson()
        )
    }
    
    static func OnNotificationFailedToPresent(notification: NotificareNotification) -> Event {
        return Event(
            type: .notificationFailedToPresent,
            payload: try! notification.toJson()
        )
    }
    
    static func OnNotificationUrlClicked(notification: NotificareNotification, url: URL) -> Event {
        return Event(
            type: .notificationUrlClicked,
            payload: [
                "notification": try! notification.toJson(),
                "url": url.absoluteString,
            ]
        )
    }
    
    static func OnActionWillExecute(notification: NotificareNotification, action: NotificareNotification.Action) -> Event {
        return Event(
            type: .actionWillExecute,
            payload: [
                "notification": try! notification.toJson(),
                "action": try! action.toJson(),
            ]
        )
    }
    
    static func OnActionExecuted(notification: NotificareNotification, action: NotificareNotification.Action) -> Event {
        return Event(
            type: .actionExecuted,
            payload: [
                "notification": try! notification.toJson(),
                "action": try! action.toJson(),
            ]
        )
    }
    
    static func OnActionNotExecuted(notification: NotificareNotification, action: NotificareNotification.Action) -> Event {
        return Event(
            type: .actionNotExecuted,
            payload: [
                "notification": try! notification.toJson(),
                "action": try! action.toJson(),
            ]
        )
    }
    
    static func OnActionFailedToExecute(notification: NotificareNotification, action: NotificareNotification.Action, error: Error?) -> Event {
        var payload: [String: Any] = [
            "notification": try! notification.toJson(),
            "action": try! action.toJson(),
        ]
        
        if let error = error {
            payload["error"] = error.localizedDescription
        }
        
        return Event(
            type: .actionFailedToExecute,
            payload: payload
        )
    }
    
    static func OnCustomActionReceived(url: URL) -> Event {
        return Event(
            type: .customActionReceived,
            payload: url.absoluteString
        )
    }
}

File: notificare_push_ui/android/src/main/kotlin/re/notifica/push/ui/flutter/NotificarePushUIPluginEventBroker.kt
package re.notifica.push.ui.flutter

import android.net.Uri
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.JSONMethodCodec
import re.notifica.models.NotificareNotification

internal object NotificarePushUIPluginEventBroker {

    private val streams: Map<Event.Type, Stream> by lazy {
        Event.Type.values().associate {
            it to Stream(it)
        }
    }

    fun register(messenger: BinaryMessenger) {
        streams.values.forEach {
            val channel = EventChannel(messenger, it.name, JSONMethodCodec.INSTANCE)
            channel.setStreamHandler(it)
        }
    }

    fun emit(event: Event) {
        Handler(Looper.getMainLooper()).post {
            streams[event.type]?.emit(event)
        }
    }


    sealed class Event {

        abstract val type: Type
        abstract val payload: Any?

        enum class Type(val id: String) {
            NOTIFICATION_WILL_PRESENT(id = "notification_will_present"),
            NOTIFICATION_PRESENTED(id = "notification_presented"),
            NOTIFICATION_FINISHED_PRESENTING(id = "notification_finished_presenting"),
            NOTIFICATION_FAILED_TO_PRESENT(id = "notification_failed_to_present"),
            NOTIFICATION_URL_CLICKED(id = "notification_url_clicked"),
            ACTION_WILL_EXECUTE(id = "action_will_execute"),
            ACTION_EXECUTED(id = "action_executed"),
            ACTION_NOT_EXECUTED(id = "action_not_executed"),
            ACTION_FAILED_TO_EXECUTE(id = "action_failed_to_execute"),
            CUSTOM_ACTION_RECEIVED(id = "custom_action_received"),
        }

        class NotificationWillPresent(
            notification: NotificareNotification
        ) : Event() {
            override val type = Type.NOTIFICATION_WILL_PRESENT
            override val payload = notification.toJson()
        }

        class NotificationPresented(
            notification: NotificareNotification
        ) : Event() {
            override val type = Type.NOTIFICATION_PRESENTED
            override val payload = notification.toJson()
        }

        class NotificationFinishedPresenting(
            notification: NotificareNotification
        ) : Event() {
            override val type = Type.NOTIFICATION_FINISHED_PRESENTING
            override val payload = notification.toJson()
        }

        class NotificationFailedToPresent(
            notification: NotificareNotification
        ) : Event() {
            override val type = Type.NOTIFICATION_FAILED_TO_PRESENT
            override val payload = notification.toJson()
        }

        class NotificationUrlClicked(
            notification: NotificareNotification,
            uri: Uri,
        ) : Event() {
            override val type = Type.NOTIFICATION_URL_CLICKED
            override val payload = mapOf(
                "notification" to notification.toJson(),
                "url" to uri.toString(),
            )
        }

        class ActionWillExecute(
            notification: NotificareNotification,
            action: NotificareNotification.Action,
        ) : Event() {
            override val type = Type.ACTION_WILL_EXECUTE
            override val payload = mapOf(
                "notification" to notification.toJson(),
                "action" to action.toJson(),
            )
        }

        class ActionExecuted(
            notification: NotificareNotification,
            action: NotificareNotification.Action,
        ) : Event() {
            override val type = Type.ACTION_EXECUTED
            override val payload = mapOf(
                "notification" to notification.toJson(),
                "action" to action.toJson(),
            )
        }

        @Suppress("unused")
        class ActionNotExecuted(
            notification: NotificareNotification,
            action: NotificareNotification.Action,
        ) : Event() {
            override val type = Type.ACTION_NOT_EXECUTED
            override val payload = mapOf(
                "notification" to notification.toJson(),
                "action" to action.toJson(),
            )
        }

        class ActionFailedToExecute(
            notification: NotificareNotification,
            action: NotificareNotification.Action,
            error: Exception?,
        ) : Event() {
            override val type = Type.ACTION_FAILED_TO_EXECUTE
            override val payload = mapOf(
                "notification" to notification.toJson(),
                "action" to action.toJson(),
                "error" to error?.message,
            )
        }

        class CustomActionReceived(
            notification: NotificareNotification,
            action: NotificareNotification.Action,
            uri: Uri,
        ) : Event() {
            override val type = Type.CUSTOM_ACTION_RECEIVED
            override val payload = mapOf(
                "notification" to notification.toJson(),
                "action" to action.toJson(),
                "uri" to uri.toString(),
            )
        }
    }


    class Stream(type: Event.Type) : EventChannel.StreamHandler {

        private var eventSink: EventChannel.EventSink? = null
        private val pendingEvents = mutableListOf<Event>()

        val name = "${NotificarePushUIPlugin.NAMESPACE}/events/${type.id}"

        fun emit(event: Event) {
            val eventSink = this.eventSink

            if (eventSink == null) {
                pendingEvents.add(event)
            } else {
                eventSink.success(event.payload)
            }
        }

        override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink?) {
            this.eventSink = eventSink

            if (eventSink != null) {
                pendingEvents.forEach(::emit)
                pendingEvents.clear()
            }
        }

        override fun onCancel(arguments: Any?) {
            this.eventSink = null
        }
    }
}
